crates/ecs/src/lib.rs

```diff
@@ -503,74 +503,355 @@ pub struct World {
+
+                // Update component_type_id_to_archetype when the the incoming
+                // component type had previously existed. Alternatively not existed.
+                let type_id_of_added_component = TypeId::of::<ComponentType>();
+
+                if let Some(archetype_indices) = self
+                    .component_typeid_to_archetype_indices
+                    .get_mut(&type_id_of_added_component)
+                {
+                    archetype_indices.insert(target_archetype_idx);
+                } else {
+                    let mut new_archetype_indices = HashSet::new();
+                    new_archetype_indices.insert(target_archetype_idx);
+                    self.component_typeid_to_archetype_indices
+                        .insert(type_id_of_added_component, new_archetype_indices);
+                }
+            }
+        }
+
+        self.update_source_archetype_after_entity_move(
+            entity,
+            source_archetype_index,
+            source_component_vec_idx,
+        )?;
+
+        Ok(())
+    }
+
+    fn remove_component_type_from_entity<ComponentType: Debug + Send + Sync + 'static>(
+        &mut self,
+        entity: Entity,
+    ) -> WorldResult<()> {
+        let source_archetype_index = *self
+            .entity_to_archetype_index
+            .get(&entity)
+            .ok_or(WorldError::EntityDoesNotExist(entity))?;
+
+        let source_component_vec_idx = *self
+            .archetypes
+            .get(source_archetype_index)
+            .ok_or(WorldError::ArchetypeDoesNotExist(source_archetype_index))?.entity_to_component_index
+            .get(&entity)
+            .expect("Entity should have a component index tied to it since it is already established to be existing within the archetype");
+
+        let (target_archetype_exists, (source_type_ids, target_type_ids)) =
+            self.find_target_archetype::<ComponentType>(entity, true)?;
+
+        // Handle if an archetype that contains only relevant components exists or not
+        match target_archetype_exists {
+            Some(target_archetype_idx) => {
+                self.move_entity_components_between_archetypes(
+                    entity,
+                    target_archetype_idx,
+                    source_type_ids,
+                )?;
+            }
+            None => {
+                let target_archetype_idx =
+                    self.move_entity_components_to_new_archetype(entity, target_type_ids)?;
+
+                // Remove component
+                let source_archetype = self
+                    .archetypes
+                    .get(source_archetype_index)
+                    .expect("Should exist since it existed in the begining of the method");
+
+                let target_archetype = self
+                    .archetypes
+                    .get(target_archetype_idx)
+                    .expect("Archetype should have been created by previous function");
```

Comment:
Can't remember. Will check and make more clear

Reply:
> Semms strange to move a value and then fetch it instantly within a function.

Yeah, but that's what you're doing but outside of the function. Inside `move_entity_components_to_new_archetype` you create an archetype, and then you return an index to it. Outside of the function (in the code highlighted above) you then  fetch the moved value (the archetype) using the index. That's mostly what I find strange, that you return an index to something you could return a reference to, only to then use that index immediately to get a reference. It seems like an unnecessary step required by the user of the function

Imo, returning an index to an element and returning a reference to an element are very similar and I think it's totally fine to return a reference to a value created and moved inside of the function.

But if there are other reasons for keeping it as an index, then it'd be more understandable, so maybe I'm missing something?

Reply:
Looking at it again, it seems you're using both the archetype index and the reference to the archetype inside the caller. So even if this function was changed to return a reference, the index is still necessary since much of the rest of the code only uses indices. 

So you can leave it as-is and resolve this 👍